projects/util-linux-2.40.2/sys-utils/renice.1
'\" t
.\"     Title: renice
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.20
.\"      Date: 2024-04-24
.\"    Manual: User Commands
.\"    Source: util-linux 2.40.2
.\"  Language: English
.\"
.TH "RENICE" "1" "2024-04-24" "util\-linux 2.40.2" "User Commands"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
renice \- alter priority of running processes
.SH "SYNOPSIS"
.sp
\fBrenice\fP [\fB\-\-priority|\-\-relative\fP] \fIpriority\fP [\fB\-g\fP|\fB\-p\fP|\fB\-u\fP] \fIidentifier\fP...
.SH "DESCRIPTION"
.sp
\fBrenice\fP alters the scheduling priority of one or more running processes. The first argument is the \fIpriority\fP value to be used. The other arguments are interpreted as process IDs (by default), process group IDs, user IDs, or user names. \fBrenice\fP\*(Aqing a process group causes all processes in the process group to have their scheduling priority altered. \fBrenice\fP\*(Aqing a user causes all processes owned by the user to have their scheduling priority altered.
.sp
If no \fB\-n\fP, \fB\-\-priority\fP or \fB\-\-relative\fP option is used, then the priority is set as \fBabsolute\fP.
.SH "OPTIONS"
.sp
\fB\-n\fP \fIpriority\fP
.RS 4
Specify the \fBabsolute\fP or \fBrelative\fP (depending on environment variable POSIXLY_CORRECT) scheduling \fIpriority\fP to be used for the process, process group, or user. Use of the option \fB\-n\fP is optional, but when used, it must be the first argument. See \fBNOTES\fP for more information.
.RE
.sp
\fB\-\-priority\fP \fIpriority\fP
.RS 4
Specify an \fBabsolute\fP scheduling \fIpriority\fP. \fIPriority\fP is set to the given value. This is the default, when no option is specified.
.RE
.sp
\fB\-\-relative\fP \fIpriority\fP
.RS 4
Specify a \fBrelative\fP scheduling \fIpriority\fP. Same as the standard POSIX \fB\-n\fP option. \fIPriority\fP gets \fIincremented/decremented\fP by the given value.
.RE
.sp
\fB\-g\fP, \fB\-\-pgrp\fP
.RS 4
Interpret the succeeding arguments as process group IDs.
.RE
.sp
\fB\-p\fP, \fB\-\-pid\fP
.RS 4
Interpret the succeeding arguments as process IDs (the default).
.RE
.sp
\fB\-u\fP, \fB\-\-user\fP
.RS 4
Interpret the succeeding arguments as usernames or UIDs.
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Display help text and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-version\fP
.RS 4
Print version and exit.
.RE
.SH "FILES"
.sp
\fI/etc/passwd\fP
.RS 4
to map user names to user IDs
.RE
.SH "NOTES"
.sp
Users other than the superuser may only alter the priority of processes they own. Furthermore, an unprivileged user can only \fIincrease\fP the "nice value" (i.e., choose a lower priority) and such changes are irreversible unless (since Linux 2.6.12) the user has a suitable "nice" resource limit (see \fBulimit\fP(1p) and \fBgetrlimit\fP(2)).
.sp
The superuser may alter the priority of any process and set the priority to any value in the range \-20 to 19. Useful priorities are: 19 (the affected processes will run only when nothing else in the system wants to), 0 (the "base" scheduling priority), anything negative (to make things go very fast).
.sp
For historical reasons in this implementation, the \fB\-n\fP option did not follow the POSIX specification. Therefore, instead of setting a \fBrelative\fP priority, it sets an \fBabsolute\fP priority by default. As this may not be desirable, this behavior can be controlled by setting the environment variable POSIXLY_CORRECT to be fully POSIX compliant. See the \fB\-n\fP option for details. See \fB\-\-relative\fP and \fB\-\-priority\fP for options that do not change behavior depending on environment variables.
.SH "HISTORY"
.sp
The \fBrenice\fP command appeared in 4.0BSD.
.SH "EXAMPLES"
.sp
The following command would change the priority of the processes with PIDs 987 and 32, plus all processes owned by the users daemon and root:
.sp
\fBrenice +1 987 \-u daemon root \-p 32\fP
.SH "SEE ALSO"
.sp
\fBnice\fP(1),
\fBchrt\fP(1),
\fBgetpriority\fP(2),
\fBsetpriority\fP(2),
\fBcredentials\fP(7),
\fBsched\fP(7)
.SH "REPORTING BUGS"
.sp
For bug reports, use the issue tracker at \c
.URL "https://github.com/util\-linux/util\-linux/issues" "" "."
.SH "AVAILABILITY"
.sp
The \fBrenice\fP command is part of the util\-linux package which can be downloaded from \c
.URL "https://www.kernel.org/pub/linux/utils/util\-linux/" "Linux Kernel Archive" "."

projects/util-linux-2.40.2/sys-utils/renice.c
/*
 * SPDX-License-Identifier: BSD-4-Clause-UC
 *
 * Copyright (c) 1983, 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

 /* 1999-02-22 Arkadiusz Miśkiewicz <misiek@pld.ORG.PL>
  * - added Native Language Support
  */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/resource.h>

#include <stdio.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include "nls.h"
#include "c.h"
#include "closestream.h"

static const char *idtype[] = {
	[PRIO_PROCESS]	= N_("process ID"),
	[PRIO_PGRP]	= N_("process group ID"),
	[PRIO_USER]	= N_("user ID"),
};

static void __attribute__((__noreturn__)) usage(void)
{
	FILE *out = stdout;
	fputs(USAGE_HEADER, out);
	fprintf(out,
	      _(" %1$s [-n|--priority|--relative] <priority> [-p|--pid] <pid>...\n"
		" %1$s [-n|--priority|--relative] <priority>  -g|--pgrp <pgid>...\n"
		" %1$s [-n|--priority|--relative] <priority>  -u|--user <user>...\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Alter the priority of running processes.\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -n <num>               specify the nice value\n"), out);
	fputs(_("                          If POSIXLY_CORRECT flag is set in environment\n"), out);
	fputs(_("                          then the priority is 'relative' to current\n"), out);
	fputs(_("                          process priority. Otherwise it is 'absolute'.\n"), out);
	fputs(_(" --priority <num>       specify the 'absolute' nice value\n"), out);
	fputs(_(" --relative <num>       specify the 'relative' nice value\n"), out);
	fputs(_(" -p, --pid              interpret arguments as process ID (default)\n"), out);
	fputs(_(" -g, --pgrp             interpret arguments as process group ID\n"), out);
	fputs(_(" -u, --user             interpret arguments as username or user ID\n"), out);
	fputs(USAGE_SEPARATOR, out);
	fprintf(out, USAGE_HELP_OPTIONS(24));
	fprintf(out, USAGE_MAN_TAIL("renice(1)"));
	exit(EXIT_SUCCESS);
}

static int getprio(const int which, const int who, int *prio)
{
	errno = 0;
	*prio = getpriority(which, who);
	if (*prio == -1 && errno) {
		warn(_("failed to get priority for %d (%s)"), who, idtype[which]);
		return -errno;
	}
	return 0;
}

static int donice(const int which, const int who, const int prio, const int relative)
{
	int oldprio, newprio;

	if (getprio(which, who, &oldprio) != 0)
		return 1;

	newprio = prio; // if not relative, set absolute priority

	if (relative)
		newprio = oldprio + prio;

	if (setpriority(which, who, newprio) < 0) {
		warn(_("failed to set priority for %d (%s)"), who, idtype[which]);
		return 1;
	}
	if (getprio(which, who, &newprio) != 0)
		return 1;
	printf(_("%d (%s) old priority %d, new priority %d\n"),
	       who, idtype[which], oldprio, newprio);
	return 0;
}

/*
 * Change the priority (the nice value) of processes
 * or groups of processes which are already running.
 */
int main(int argc, char **argv)
{
	int which = PRIO_PROCESS;
	int who = 0, prio, errs = 0;
	int relative = 0;
	char *endptr = NULL;

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	close_stdout_atexit();

	argc--;
	argv++;

	if (argc == 1) {
		if (strcmp(*argv, "-h") == 0 ||
		    strcmp(*argv, "--help") == 0)
			usage();

		if (strcmp(*argv, "-v") == 0 ||
		    strcmp(*argv, "-V") == 0 ||
		    strcmp(*argv, "--version") == 0)
			print_version(EXIT_SUCCESS);
	}

	if (*argv)
	{
		if (strcmp(*argv, "-n") == 0)
		{
			// Fully conform to posix only if POSIXLY_CORRECT is
			// set in the environment. If not, use the absolute
			// value as it's been used (incorrectly) since 2009
			relative = getenv("POSIXLY_CORRECT") == NULL ? 0 : 1;
			argc--;
			argv++;
		}
		else if (strcmp(*argv, "--relative") == 0)
		{
			relative = 1;
			argc--;
			argv++;
		}
		else if (strcmp(*argv, "--priority") == 0) {
			relative = 0;
			argc--;
			argv++;
		}
	}

	if (argc < 2 || !*argv) {
		warnx(_("not enough arguments"));
		errtryhelp(EXIT_FAILURE);
	}

	prio = strtol(*argv, &endptr, 10);
	if (*endptr) {
		warnx(_("invalid priority '%s'"), *argv);
		errtryhelp(EXIT_FAILURE);
	}
	argc--;
	argv++;

	for (; argc > 0; argc--, argv++) {
		if (strcmp(*argv, "-g") == 0 || strcmp(*argv, "--pgrp") == 0) {
			which = PRIO_PGRP;
			continue;
		}
		if (strcmp(*argv, "-u") == 0 || strcmp(*argv, "--user") == 0) {
			which = PRIO_USER;
			continue;
		}
		if (strcmp(*argv, "-p") == 0 || strcmp(*argv, "--pid") == 0) {
			which = PRIO_PROCESS;
			continue;
		}
		if (which == PRIO_USER) {
			struct passwd *pwd = getpwnam(*argv);

			if (pwd != NULL)
				who = pwd->pw_uid;
			else
				who = strtol(*argv, &endptr, 10);
			if (who < 0 || *endptr) {
				warnx(_("unknown user %s"), *argv);
				errs = 1;
				continue;
			}
		} else {
			who = strtol(*argv, &endptr, 10);
			if (who < 0 || *endptr) {
				/* TRANSLATORS: The first %s is one of the above
				 * three ID names. Read: "bad value for %s: %s" */
				warnx(_("bad %s value: %s"), idtype[which], *argv);
				errs = 1;
				continue;
			}
		}
		errs |= donice(which, who, prio, relative);
	}
	return errs != 0 ? EXIT_FAILURE : EXIT_SUCCESS;
}
